File: benchmarks/llm/gpt2_compare.py
"""Diff R2's GPT-2 token ids against HuggingFace's, on the same vocabulary.

Run the Rust half first:

    R2_GPT2_JSON=<path to gpt2_tokenizer.json> \
      cargo run --release -p r2-tensor --example gpt2_compare > r2_ids.txt
    python benchmarks/llm/gpt2_compare.py r2_ids.txt <path to gpt2_tokenizer.json>

Every other claim about matching GPT-2 is an argument from the published
algorithm. This is the only evidence: one vocabulary, one input, two
implementations, compared id for id.
"""

import sys
import time

from tokenizers import Tokenizer

CASES = [
    "hello world", " hello", "hello", "don't stop",
    "It's 42 apples, isn't it?", "a   b", "  leading", "trailing  ",
    "x=1;y=2", "The quick brown fox jumps over the lazy dog.",
    "1234567890", "CamelCaseIdentifier", "snake_case_name",
    "tabs\tand\nnewlines", "日本語のテキスト", "emoji 🙂 and 🎉 mixed",
    "café naïve résumé", "  ", "a", "",
]


def main():
    r2_path = sys.argv[1] if len(sys.argv) > 1 else "r2_ids.txt"
    json_path = sys.argv[2] if len(sys.argv) > 2 else "gpt2_tokenizer.json"

    hf = Tokenizer.from_file(json_path)
    r2 = {}
    r2_throughput = None
    for line in open(r2_path, encoding="utf-8"):
        line = line.rstrip("\n")
        if line.startswith("CASE "):
            _, idx, rest = line.split(" ", 2)
            r2[int(idx)] = ([] if not rest or rest == "" else
                            [int(x) for x in rest.split(",") if x != ""])
        elif line.startswith("THROUGHPUT"):
            r2_throughput = dict(
                kv.split("=") for kv in line.split()[1:])

    print(f"{'#':>3} {'case':<34} {'HF':>6} {'R2':>6}  verdict")
    print("-" * 72)
    agree = 0
    for i, case in enumerate(CASES):
        want = hf.encode(case).ids
        got = r2.get(i)
        if got is None:
            print(f"{i:>3} {case!r:<34} {len(want):>6} {'-':>6}  MISSING")
            continue
        ok = want == got
        agree += ok
        label = case.replace("\n", "\\n").replace("\t", "\\t")
        print(f"{i:>3} {label!r:<34} {len(want):>6} {len(got):>6}  "
              f"{'match' if ok else 'DIFFER'}")
        if not ok:
            print(f"      HF: {want}")
            print(f"      R2: {got}")
            print(f"      HF pieces: {[hf.decode([t]) for t in want]}")

    print("-" * 72)
    print(f"{agree}/{len(CASES)} cases identical to HuggingFace")

    # Throughput on the same body the Rust side timed.
    body = " ".join(CASES) * 2000
    t0 = time.perf_counter()
    ids = hf.encode(body).ids
    secs = time.perf_counter() - t0
    hf_mbps = len(body.encode()) / secs / 1e6
    print(f"\nthroughput on {len(body.encode())/1e6:.1f} MB")
    print(f"  HuggingFace {hf_mbps:>8.2f} MB/s  ({len(ids)} tokens)")
    if r2_throughput:
        r2_mbps = float(r2_throughput["mbps"])
        print(f"  Ardon-R2    {r2_mbps:>8.2f} MB/s  "
              f"({r2_throughput['tokens']} tokens)")
        print(f"  HF is {hf_mbps / r2_mbps:.1f}x faster")


if __name__ == "__main__":
    main()

File: crates/r2-tensor/examples/gpt2_compare.rs
//! Load GPT-2's real `tokenizer.json` and dump R2's token ids, so they can
//! be diffed against HuggingFace's on the SAME file and the SAME input.
//!
//! This is the check that settles whether R2 "matches the GPT-2 standard".
//! Everything else — the byte alphabet, the merge ranks, the pre-tokenizer
//! regex — is an argument from the published algorithm. Only feeding both
//! implementations one vocabulary and comparing ids is evidence.
//!
//!     R2_GPT2_JSON=gpt2_tokenizer.json \
//!       cargo run --release -p r2-tensor --example gpt2_compare
//!
//! Prints one line per case: `CASE <n> <ids...>`, plus a throughput figure
//! on a larger body of text. `gpt2_compare.py` reads it and diffs.

use r2_tensor::tokenizer::Tokenizer;

/// The cases that separate a correct ByteLevel BPE from a plausible one.
/// Every entry exercises a specific clause of GPT-2's pre-tokenizer regex.
const CASES: &[&str] = &[
    "hello world",
    " hello",
    "hello",
    "don't stop",
    "It's 42 apples, isn't it?",
    "a   b",                       // \s+(?!\S): run of spaces
    "  leading",                   // leading whitespace run
    "trailing  ",                  // trailing whitespace run
    "x=1;y=2",                     // punctuation runs
    "The quick brown fox jumps over the lazy dog.",
    "1234567890",
    "CamelCaseIdentifier",
    "snake_case_name",
    "tabs\tand\nnewlines",
    "日本語のテキスト",              // CJK, multi-byte
    "emoji 🙂 and 🎉 mixed",        // 4-byte UTF-8
    "café naïve résumé",           // Latin-1 range
    "  ",                          // whitespace only
    "a",
    "",
];

fn main() {
    let path = std::env::var("R2_GPT2_JSON")
        .unwrap_or_else(|_| "gpt2_tokenizer.json".into());
    let src = std::fs::read_to_string(&path).unwrap_or_else(|e| {
        eprintln!("cannot read {path}: {e}");
        std::process::exit(1);
    });
    let tok = Tokenizer::from_tokenizer_json(&src).unwrap_or_else(|e| {
        eprintln!("cannot parse {path}: {e}");
        std::process::exit(1);
    });
    eprintln!("loaded {} tokens, encoding {:?}", tok.vocab_size(), tok.encoding());

    for (i, case) in CASES.iter().enumerate() {
        match tok.encode(case) {
            Ok(ids) => {
                let s: Vec<String> = ids.iter().map(|i| i.to_string()).collect();
                println!("CASE {i} {}", s.join(","));
            }
            Err(e) => println!("CASE {i} ERROR {e}"),
        }
    }

    // SCALING. A prior analysis of this code measured encode as O(n^2)
    // (3,808 bytes in 0.94 s, extrapolating to ~18 hours per MB) because BPE
    // ran over the whole document instead of per pre-token. If that is
    // fixed, time must roughly DOUBLE as the input doubles, not quadruple.
    println!("SCALING");
    let unit = CASES.join(" ");
    for k in [1usize, 2, 4, 8, 16, 32, 64] {
        let body = unit.repeat(k * 8);
        let t0 = std::time::Instant::now();
        let ids = tok.encode(&body).expect("encode");
        let secs = t0.elapsed().as_secs_f64();
        println!("  bytes={:>8} secs={:.5} mbps={:>6.2} tokens={}",
                 body.len(), secs, body.len() as f64 / secs / 1e6, ids.len());
    }

    // PHASE SPLIT: how much of encode is the pre-tokenizer, and how much
    // is the merge loop? Optimising the wrong half is the standing risk.
    {
        let body = CASES.join(" ").repeat(2000);
        let reps = 5;
        let t0 = std::time::Instant::now();
        let mut words = 0usize;
        for _ in 0..reps { words = r2_tensor::bpe::pretokenize(&body).len(); }
        let pre = t0.elapsed().as_secs_f64() / reps as f64;
        let t0 = std::time::Instant::now();
        for _ in 0..reps { std::hint::black_box(tok.encode(&body).unwrap().len()); }
        let full = t0.elapsed().as_secs_f64() / reps as f64;
        println!("PHASES bytes={} words={words} pretokenize={:.4}s ({:.1}%) rest={:.4}s ({:.1}%) total={:.4}s",
                 body.len(), pre, pre / full * 100.0,
                 full - pre, (full - pre) / full * 100.0, full);
    }

    // Throughput on a larger body, and the round-trip that makes it usable.
    let body = CASES.join(" ").repeat(2000);
    let t0 = std::time::Instant::now();
    let ids = tok.encode(&body).expect("encode");
    let secs = t0.elapsed().as_secs_f64();
    println!("THROUGHPUT bytes={} tokens={} secs={:.4} mbps={:.2}",
             body.len(), ids.len(), secs, body.len() as f64 / secs / 1e6);
    println!("ROUNDTRIP {}", if tok.decode(&ids) == body { "EXACT" } else { "MISMATCH" });
}
